src/ams.py

```diff
@@ -0,0 +1,187 @@
+GROUP_NAME = "ams"
+
+
+class AMS:
+    """Class for handling AMS configurations."""
+
+    def __init__(self, charm: ops.CharmBase):
+        self._sc = snap.SnapCache()
+        self._charm = charm
+
+    def restart(self):
+        """Restart AMS Snap."""
+        self._sc["ams"].restart()
+
+    def remove(self):
+        """Remove AMS Snap."""
+        self._sc["ams"]._remove()
+
+    def install(self):
+        """Install AMS including its Snap."""
+        try:
+            res = self._charm.model.resources.fetch("ams-snap")
+        except ops.ModelError:
+            res = None
+
+        # FIXME: Install the ams snap from a resource until we make the
+        # snaps in the snap store unlisted
+        if res is not None and res.stat().st_size:
+            snap.install_local(res, classic=False, dangerous=True)
+        else:
+            self._charm.unit.status = BlockedStatus("Waiting for AMS snap resource")
+            return
+
+        ams_snap = self._sc["ams"]
+        ams_snap.connect(plug="daemon-notify", slot="core:daemon-notify")
+        ams_snap.alias("amc", "amc")
+
+        passwd.add_group(GROUP_NAME)
+        passwd.add_user_to_group("ubuntu", GROUP_NAME)
+        self._create_systemd_drop_in()
+
+    def _create_systemd_drop_in(self):
+        tenv = Environment(loader=FileSystemLoader("templates"))
+        template = tenv.get_template("10-ams-unix-socket-chown.conf.j2")
+        rendered_content = template.render(
+            {
+                "group": GROUP_NAME,
+            }
+        )
+        SERVICE_DROP_IN_PATH.parent.mkdir(parents=True, exist_ok=True)
+        SERVICE_DROP_IN_PATH.write_text(rendered_content)
+        systemd.daemon_reload()
+
+    # TODO: remove this function to get snap from SnapCache()['ams'] after the
+    # snap is made publicly available in the snap store
+    def _get_snap(self) -> dict | None:
+        snaps = self._sc._snap_client.get_installed_snaps()
+        for installed_snap in snaps:
+            if installed_snap["name"] == SNAP_NAME:
+                return installed_snap
+        return None
+
+    @property
+    def version(self) -> str:
+        """Return AMS version."""
+        _snap = self._get_snap()
+        if not _snap:
+            raise snap.SnapNotFoundError(SNAP_NAME)
+        return _snap["version"]
+
+    @property
+    def installed(self) -> bool:
+        """Check if AMS is installed."""
+        _snap = self._get_snap()
+        if not _snap:
+            return False
+        return True
+
+    @staticmethod
+    def get_ip_for_interface(interface):
+        """Return the ip address associated to the given interface."""
+        addresses = netifaces.ifaddresses(interface)
+        if netifaces.AF_INET not in addresses or "addr" not in addresses[netifaces.AF_INET][0]:
+            raise Exception("No IP associated to requested device")
+        return addresses[netifaces.AF_INET][0]["addr"]
+
+    def configure(
```

Comment:
Where is this called from?

Reply:
The updated PR has this called in the config_changed hook.